src/router.rs

```diff
@@ -5,27 +5,57 @@ use crate::parser::parse;
 use crate::schema::Schema;
 use crate::semantics::{FieldCounter, Validate};
 use std::collections::{BTreeMap, HashMap};
+use std::ops::Deref;
 use uuid::Uuid;
 
 #[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
 struct MatcherKey(usize, Uuid);
 
+#[derive(Debug)]
+enum SchemaOwnedOrRef<'a> {
+    Ref(&'a Schema),
+    Owned(Box<Schema>),
+}
+
+impl Deref for SchemaOwnedOrRef<'_> {
+    type Target = Schema;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            Self::Ref(s) => s,
+            Self::Owned(s) => s,
+        }
+    }
+}
+
 #[derive(Debug)]
 pub struct Router<'a> {
-    schema: &'a Schema,
+    schema: SchemaOwnedOrRef<'a>,
     matchers: BTreeMap<MatcherKey, Expression>,
     pub fields: HashMap<String, usize>,
 }
 
 impl<'a> Router<'a> {
     pub fn new(schema: &'a Schema) -> Self {
         Self {
-            schema,
+            schema: SchemaOwnedOrRef::Ref(schema),
+            matchers: BTreeMap::new(),
+            fields: HashMap::new(),
+        }
+    }
+
+    pub fn new_owning(schema: Schema) -> Self {
```

Comment:
NIT: might be a little bit weird using `owning` here? I might be wrong.
```suggestion
    pub fn new_owned(schema: Schema) -> Self {
```

Reply:
new "owning" was meant as in: this constructor generates a new "owning" router, i.e. one that owns a Schema